onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -227,13 +227,7 @@ Status CUDAExecutionProvider::CopyTensor(const Tensor& src, Tensor& dst, int exe
       CUDA_RETURN_IF_ERROR(cudaMemcpy(dst_data, src_data, bytes, cudaMemcpyHostToDevice));
     }
   } else if (strcmp(src.Location().name, CUDA) == 0) {
-    if (strcmp(dst.Location().name, CUDA_PINNED) == 0) {
-      // copying from GPU to pinned memory, this is non-blocking
-      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(dst_data, src_data, bytes, cudaMemcpyDeviceToHost, streams_[exec_queue_id]));
```

Comment:
>cudaMemcpyAsync [](start = 27, length = 15)

Shouldn't it be guarded by CUDAFence?

Reply:
CUDAFence only sync between default stream (which kernel running) with copy stream. which means the kernel execution always after the copy complete. however this is copy to CPU. The code before kernel launch need to access the CPU memory.

Reply:
just notice there's a also cudaEventSynchronize, still need to investigate more.

---
In reply to: [291484105](https://github.com/microsoft/onnxruntime/pull/1182#discussion_r291484105) [](ancestors = 291484105)

Reply:
Note that from document cuda event seems to only have unsignaled and signaled states. So my guess is that our CUDAFence implementation should not be reusing events before it is signaled. Considering following cases in timeline:

t0 (CPU): unsignal event E, and queue command G1 on GPU
t1 (CPU) unsignal event E, and queue command G2 on GPU
t2 (GPU) signal event E to mark command G1 completed
t3 (GPU) signal event E to mark command G2 completed

Then we may have a bug when CPU checks if G2 is completed between t2 and t3, as the event would be signaled at t2, but CPU assumed it means G2 complete.

If this is the bug, we need to have a pool of event in each CUDAFence, and when CPU found signaled event it would be put back to pool for reuse. CPU creates a new event if previous event has not been signaled yet. Then the timeline won't have issue:

t0 (CPU): unsignal event E1, and queue command G1 on GPU
t1 (CPU) since E1 is not signaled, create/reuse from pool with unsignal event E2, and queue command G2 on GPU
t2 (GPU) signal event E1 to mark command G1 completed
t3 (GPU) signal event E2 to mark command G2 completed



---
In reply to: [291780135](https://github.com/microsoft/onnxruntime/pull/1182#discussion_r291780135) [](ancestors = 291780135,291484105)